pkg/portal/graphql/app_mutation.go

```diff
@@ -287,11 +287,44 @@ var _ = registerMutationField(
 				if err != nil {
 					return nil, err
 				}
+
 				appConfigYAML, err := yaml.JSONToYAML(appConfigJSON)
 				if err != nil {
 					return nil, err
 				}
 
+				appConfig, err := config.Parse([]byte(appConfigYAML))
+				if err != nil {
+					return nil, err
+				}
+
+				if appConfig.HTTP.PublicOrigin != originalAppConfig.HTTP.PublicOrigin {
+					validOrigin := false
+
+					for _, appHostSuffix := range gqlCtx.AppHostSuffixes {
+						appHost := appID + appHostSuffix
+						if appConfig.HTTP.PublicOrigin == appHost {
+							validOrigin = true
+							break
+						}
+					}
+
+					availableDomains, err := gqlCtx.DomainService.ListDomains(appID)
+					if err != nil {
+						return nil, err
+					}
+
+					for _, domain := range availableDomains {
+						if domain.Domain == appConfig.HTTP.PublicOrigin {
+							validOrigin = true
+							break
+						}
+					}
+					if !validOrigin {
+						return nil, errors.New("public origin must be one of the available domains")
+					}
+				}
+
```

Comment:
- In pkg/portal/appresource/manager.go, inject appHostSuffixes in context
- In pkg/portal/appresource/manager.go, inject domain service in context. The domain service should be able to list custom domains of a project.
- In pkg/lib/config/configsource/resources.go, update AuthgearYAML UpdateResources to use the newly injected values to perform domain validation.

Reply:
Updated 🙏